src/Accelerators/NNPA/Runtime/zDNNExtension/zDNNExtension.c

```diff
@@ -103,42 +103,145 @@ static zdnn_status allocZTensorChunk(
+  case (ZDNN_1D):
+    // shape (a) <- dims4-1 (1, 1, 1, a)
+    if (axis == 3)
+      preTransDesc->dim1 = chunkSize;
+    break;
+  case (ZDNN_2D):
+    // shape (a, b) -> dims4-1 (1, 1, a, b)
+    if (axis == 2)
+      preTransDesc->dim2 = chunkSize;
+    if (axis == 3)
+      preTransDesc->dim1 = chunkSize;
+    break;
+  case (ZDNN_2DS):
+    // shape (a, b) -> dims4-1 (a, 1, 1, b)
+    if (axis == 0)
+      preTransDesc->dim2 = chunkSize;
+    if (axis == 3)
+      preTransDesc->dim1 = chunkSize;
+    break;
+  case (ZDNN_3D):
+    // shape (a, b, c) -> dims4-1 (1, a, b, c)
+    if (axis == 1)
+      preTransDesc->dim3 = chunkSize;
+    if (axis == 2)
+      preTransDesc->dim2 = chunkSize;
+    if (axis == 3)
+      preTransDesc->dim1 = chunkSize;
+    break;
+  case (ZDNN_3DS):
+    // shape (a, b, c) -> dims4-1 (a, 1, b, c)
+    if (axis == 0)
+      preTransDesc->dim3 = chunkSize;
+    if (axis == 2)
+      preTransDesc->dim2 = chunkSize;
+    if (axis == 3)
+      preTransDesc->dim1 = chunkSize;
+    break;
+  case (ZDNN_4D):
+  case (ZDNN_NHWC):
+  case (ZDNN_HWCK):
+    // shape (a, b, c, d) -> dims4-1 (a, b, c, d)
+    // shape (n, h, w, c) -> dims4-1 (n, h, w, c)
+    if (axis == 0)
+      preTransDesc->dim4 = chunkSize;
+    if (axis == 1)
+      preTransDesc->dim3 = chunkSize;
+    if (axis == 2)
+      preTransDesc->dim2 = chunkSize;
+    if (axis == 3)
+      preTransDesc->dim1 = chunkSize;
+    break;
+  case (ZDNN_4DS):
+    // ZDNN_4DS is used exclusively as RNN output
+    // shape (a, b, c, d)  -> ZDNN_NHWC
+    //   when b = 1 (uni-dir)     -> dims4-1 (a, 1, c, d)
+    //   otherwise (bi-dir, etc.) -> dims4-1 (a, 1, c, b * PADDED(d))
+    if (axis == 0)
+      preTransDesc->dim4 = chunkSize;
+    if (axis == 2)
+      preTransDesc->dim2 = chunkSize;
+    if (axis == 3)
+      preTransDesc->dim1 = chunkSize;
+    break;
+  case (ZDNN_NCHW):
+    // shape (n, c, h, w) -> dims4-1 (n, h, w, c)
+    if (axis == 0)
+      preTransDesc->dim4 = chunkSize;
+    if (axis == 1)
+      preTransDesc->dim2 = chunkSize;
+    if (axis == 2)
+      preTransDesc->dim1 = chunkSize;
+    if (axis == 3)
+      preTransDesc->dim3 = chunkSize;
+    break;
+  default:
+    break;
+  }
 
   // Copy a transformed desc.
   zdnn_status status = zdnn_generate_transformed_desc(preTransDesc, transDesc);
   if (status != ZDNN_OK)
     return status;
 
-  // Init a zTensor with malloc.
-  return zdnn_init_ztensor_with_malloc(preTransDesc, transDesc, chunkZTensor);
+  if (splitInfo->reuseOrigBuffer) {
+    zdnn_init_ztensor(preTransDesc, transDesc, chunkZTensor);
+    // Set a buffer size for the chunk.
+    chunkZTensor->buffer_size = zdnn_getsize_ztensor(transDesc);
+    // Set a buffer for the chunk.
+    uint64_t reuseBufferOffset;
+    if (chunkID == splitInfo->numOfChunks - 1) {
+      assert((origZTensor->buffer_size - chunkZTensor->buffer_size) %
+                 (splitInfo->numOfChunks - 1) ==
+             0);
+      reuseBufferOffset =
```

Comment:
would't the last chunk still start at a `0 mod 4k` page? Maybe you can provide a comment that provide insight.